phac-nml/irida-next
# Namespace group link user access

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in functionality for allowing access to namespaces which have been shared with groups in which the user has a membership via `NamespaceGroupLink`s. 

Addresses #221 

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

**`Group 1` shared with `Group 2`**
![image](https://github.com/phac-nml/irida-next/assets/25466211/a43564c5-d0eb-48e3-8d41-f741d80c1002)

**`Group 2` shared with `Group 3`**
![image](https://github.com/phac-nml/irida-next/assets/25466211/e2e24ccf-1732-4781-8d41-3f3e274ff0cd)

**`user2@email.com` member of `Group 2` with access to `Group 1`**

![image](https://github.com/phac-nml/irida-next/assets/25466211/0a7ca9e8-3342-413f-a172-04a605ac5ae2)

**`user2@email.com` member of `Group 2` with no access to `Group 3`**

![image](https://github.com/phac-nml/irida-next/assets/25466211/f2b42b9e-2e2f-4060-80a6-263a03401668)


## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Login as a user
2. Create 3 groups (Group A, Group B, Group C)
3. Assign a user to Group A
4. Assign a different user to Group C
5. Go to the `Groups` tab of the `Members` page for Group A
6. Invite `Group B` and set any group access level
7. Go to the `Groups` tab of the `Members` page for Group B
8. Invite `Group C` and set any access level
9. Logout and sign in as the user added above to `Group C`
10. Visit the URL for `Group B` and the user should have access to `Group B`
11. Visit the URL for `Group A` and the user should have no access to `Group A`
12. Logout and sign in as the user added above to `Group A`
13. Visit the URL for `Group B` and the user should have no access to `Group B`
14. Visit the URL for `Group C` and the user should have no access to `Group B`
15. Logout and sign in as the user added above to `Group B`
16. Visit the URL for `Group A` and the user should have access to `Group A`
17. Visit the URL for `Group C` and the user should have no access to `Group C`
18. Using the above as a template add projects to these groups and ensure the access from via `NamespaceGroupLinks` remains intact

**Note the UI has not been updated to list groups and projects which are shared via `NamespaceGroupLink`s therefore to check access you need to visit the URL for groups/projects**

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
